state/prune.go

```diff
@@ -0,0 +1,219 @@
+	var age interface{}
+
+	if p.timeUnit == NanoSeconds {
+		age = t.UnixNano()
+	} else {
+		age = t
+	}
+
+	iter := p.coll.Find(bson.D{
+		{"model-uuid", p.st.modelUUID()},
+		{p.ageField, bson.M{"$lt": age}},
+	}).Select(bson.M{"_id": 1}).Iter()
+
+	modelName, err := p.st.modelName()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logTemplate := fmt.Sprintf("%s age pruning (%s): %%d rows deleted", p.coll.Name, modelName)
+	deleted, err := p.deleteInBatches(iter, logTemplate, noEarlyFinish)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if deleted > 0 {
+		logger.Infof("%s age pruning (%s): %d rows deleted", p.coll.Name, modelName, deleted)
+	}
+	return nil
+}
+
+func (p *collectionPruner) pruneBySize() error {
+	if !p.st.isController() {
+		// Only prune by size in the controller. Otherwise we might
+		// find that multiple pruners are trying to delete the latest
+		// 1000 rows and end up with more deleted than we expect.
+		return nil
+	}
+	if p.maxSize == 0 {
+		return nil
+	}
+	// Collection Size
+	collMB, err := getCollectionMB(p.coll)
+	if err != nil {
+		return errors.Annotate(err, "retrieving collection size")
+	}
+	if collMB <= p.maxSize {
+		return nil
+	}
+	// TODO(perrito666) explore if there would be any beneffit from having the
+	// size limit be per model
+	count, err := p.coll.Count()
+	if err == mgo.ErrNotFound || count <= 0 {
+		return nil
+	}
+	if err != nil {
+		return errors.Annotatef(err, "counting %s records", p.coll.Name)
+	}
+	// We are making the assumption that status sizes can be averaged for
+	// large numbers and we will get a reasonable approach on the size.
+	// Note: Capped collections are not used for this because they, currently
+	// at least, lack a way to be resized and the size is expected to change
+	// as real life data of the history usage is gathered.
+	sizePerStatus := float64(collMB) / float64(count)
+	if sizePerStatus == 0 {
+		return fmt.Errorf("unexpected result calculating %s entry size", p.coll.Name)
+	}
+	toDelete := int(float64(collMB-p.maxSize) / sizePerStatus)
+
+	iter := p.coll.Find(nil).Sort(p.ageField).Limit(toDelete).Select(bson.M{"_id": 1}).Iter()
+
+	template := fmt.Sprintf("%s size pruning: deleted %%d of %d (estimated)", p.coll.Name, toDelete)
+	deleted, err := p.deleteInBatches(iter, template, func() (bool, error) {
+		// Check that we still need to delete more
+		collMB, err := getCollectionMB(p.coll)
+		if err != nil {
+			return false, errors.Annotatef(err, "retrieving %s collection size", p.coll.Name)
+		}
+		if collMB <= p.maxSize {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	logger.Infof("%s size pruning finished: %d rows deleted", p.coll.Name, deleted)
+
+	return nil
+}
+
+func (p *collectionPruner) deleteInBatches(iter *mgo.Iter, logTemplate string, shouldStop doneCheck) (int, error) {
+	var doc bson.M
+	chunk := p.coll.Bulk()
+	chunkSize := 0
+
+	lastUpdate := time.Now()
```

Comment:
This was a holdover from when I was "manually" testing this by running the unit test with many rows and wanted to see the logging. It should probably change back to using `p.st.clock`. (In theory that would let us test the progress logging, although don't think that's really worth doing.)